hnix-store-remote/tests/NixDaemon.hs

```diff
@@ -0,0 +1,291 @@
+import qualified Data.Vector                 as V
+import           System.Directory
+import           System.IO.Temp
+import qualified System.IO                   as IO (hGetContents, hPutStr, openFile)
+import qualified System.Process              as P
+import           System.Posix.User           as U
+import           System.Linux.Namespaces     as NS
+import           Test.Tasty                  as T
+import           Test.Tasty.Hspec            (Spec, HasCallStack, describe, context)
+import qualified Test.Tasty.Hspec            as Hspec
+import           Test.Hspec.Expectations.Lifted
+import qualified Test.Tasty.HUnit            as HU
+import           Test.Tasty.QuickCheck
+import           Text.Read                   (readMaybe)
+
+import           System.FilePath
+
+import           System.Nix.Build
+import           System.Nix.Hash
+import           System.Nix.StorePath
+import           System.Nix.ReadonlyStore
+import           System.Nix.Nar
+import qualified System.Nix.ValidPath        as VP
+import           System.Nix.Store.Remote
+import           System.Nix.Store.Remote.Logger
+import           System.Nix.Store.Remote.Types
+import           System.Nix.Store.Remote.Protocol
+import           System.Nix.Store.Remote.Util
+
+
+createProcessEnv :: FilePath -> String -> [String] -> IO P.ProcessHandle
+createProcessEnv fp proc args =do
+  (_, _, _, ph) <- P.createProcess (P.proc proc args) { P.cwd = Just $ fp
+                                                      , P.env = Just $ mockedEnv fp }
+  return ph
+
+mockedEnv :: FilePath -> [(String, FilePath)]
+mockedEnv fp = map (\(a, b) -> (a, b)) [
+    ("NIX_STORE_DIR", fp </> "store")
+  , ("NIX_LOCALSTATE_DIR", fp </> "var")
+  , ("NIX_LOG_DIR", fp </> "var" </> "log")
+  , ("NIX_STATE_DIR", fp </> "var" </> "nix")
+  , ("NIX_CONF_DIR", fp </> "etc")
+--  , ("NIX_REMOTE", "daemon")
+  ]
+
+waitSocket :: FilePath -> Int -> IO ()
+waitSocket fp 0 = fail "No socket"
+waitSocket fp x = do
+  ex <- doesFileExist fp
+  case ex of
+    True -> return ()
+    False -> threadDelay 100000 >> waitSocket fp (x - 1)
+
+writeConf fp = do
+  writeFile fp $ unlines [
+      "build-users-group = "
+    , "trusted-users = root"
+    , "allowed-users = *"
+    , "fsync-metadata = false"
+    ]
+
+{-
+ - we run in user namespace as root but groups are failed
+ - => build-users-group has to be empty but we still
+ - get an error (maybe older nix-daemon)
+ -
+uid=0(root) gid=65534(nobody) groups=65534(nobody)
+
+drwxr-xr-x 3 0 65534 60 Nov 29 05:53 store
+
+accepted connection from pid 22959, user root (trusted)
+error: changing ownership of path '/run/user/1000/test-nix-store-06b0d249e5616122/store': Invalid argument
+-}
+
+startDaemon :: FilePath -> IO (P.ProcessHandle, MonadStore a -> IO (Either String a, [Logger]))
+startDaemon fp = do
+  writeConf (fp </> "etc" </> "nix.conf")
+  p <- createProcessEnv fp "nix-daemon" []
+  writeFile "/tmp/spawned" $ show fp
```

Comment:
Possible to use `withTempFileSource` instead of hard coding `/tmp`?

Reply:
That's debug print as well, removed.